File: 1140-stone-game-ii/1140-stone-game-ii.py
class Solution:
    def stoneGameII(self, piles: list[int]) -> int:
        n = len(piles)
        
        # Initialize memoization table with 0s
        dp = [[0] * (n + 1) for _ in range(n)]
        
        # Build the suffix sum array from right to left
        suffix = [0] * (n + 1)
        for i in range(n - 1, -1, -1):
            suffix[i] = suffix[i + 1] + piles[i]
            
        def solve(i: int, m: int) -> int:
            # Base case: No piles left
            if i >= n:
                return 0
            
            # If the current player can take all remaining piles
            if 2 * m >= n - i:
                return suffix[i]
            
            # Return cached result if available
            if dp[i][m] != 0:
                return dp[i][m]
                
            best = 0
            # Explore all possible moves for taking 'x' piles
            for x in range(1, 2 * m + 1):
                next_m = max(m, x)
                opponent = solve(i + x, next_m)
                current = suffix[i] - opponent
                best = max(best, current)
                
            dp[i][m] = best
            return best

        return solve(0, 1)

File: 1140-stone-game-ii/1140-stone-game-ii.java
class Solution {

    private int[][] dp;
    private int[] suffix;
    private int n;

    public int stoneGameII(int[] piles) {
        
        n = piles.length;

        dp = new int[n][n + 1];
        suffix = new int[n + 1];

        // suffix[i] = sum of piles[i ... n - 1];
        for (int i = n - 1; i >= 0; i--)
            suffix[i] = suffix[i + 1] + piles[i];
        
        return solve(0, 1);
    }

    private int solve(int i, int m) {

        // No piles left
        if (i >= n)
            return 0;
        
        // If we can take all remaining piles
        if (2 * m >= n - i)
            return suffix[i];
        
        if (dp[i][m] != 0)
            return dp[i][m];
        
        int best = 0;

        for (int x = 1; x <= 2 * m; x++) {

            int nextM = Math.max(m,x);

            int opponent = solve(i + x, nextM);

            int current = suffix[i] - opponent;

            best = Math.max(best, current);
        }

        return dp[i][m] = best;
    }
}